datafusion/physical-expr/src/expressions/in_list.rs

```diff
@@ -209,26 +183,109 @@ macro_rules! make_contains_primitive {
     }};
 }
 
-macro_rules! set_contains_with_negated {
-    ($ARRAY:expr, $LIST_VALUES:expr, $NEGATED:expr) => {{
-        if $NEGATED {
-            return Ok(ColumnarValue::Array(Arc::new(
+macro_rules! set_contains_for_float {
+    ($ARRAY:expr, $SET_VALUES:expr, $SCALAR_VALUE:ident, $NEGATED:expr, $PHY_TYPE:ty) => {{
+        let contains_null = $SET_VALUES.iter().any(|s| s.is_null());
+        let bool_array = if $NEGATED {
+            // Not in
+            if contains_null {
                 $ARRAY
                     .iter()
-                    .map(|x| x.map(|v| !$LIST_VALUES.contains(&v.try_into().unwrap())))
-                    .collect::<BooleanArray>(),
-            )));
+                    .map(|vop| {
+                        match vop.map(|v| !$SET_VALUES.contains(&v.try_into().unwrap())) {
```

Comment:
The f32 is not implemented the `eq` trait, we can't use the `set::<f32>.contains`

Reply:
We just use the `set::<Scalarvalue::Float32/Float64>

Reply:
Elsewhere in DataFusion (including in `ScalarValue`) we use `ordered_float` to compare floating point numbers

It might be possible  to use `set::<OrderedFloat<f32>>`, which would be more space efficient (fewer bytes than `ScalarValue`) as well as faster (as the comparison doens't have to dispatch on the type each time)


https://github.com/apache/arrow-datafusion/blob/88b88d4360054a85982987aa07b3f3afd2db7d70/datafusion/common/src/scalar.rs#L33

Reply:
@alamb  I will optimize this with follow-up pr with issue https://github.com/apache/arrow-datafusion/issues/2831